substrate/frame/election-provider-support/src/lib.rs

```diff
@@ -361,72 +379,87 @@ pub trait ElectionDataProvider {
 	/// Clear all voters and targets.
 	#[cfg(any(feature = "runtime-benchmarks", test))]
 	fn clear() {}
+
+	#[cfg(any(feature = "runtime-benchmarks", test))]
+	fn set_desired_targets(_count: u32) {}
 }
 
-/// Base trait for types that can provide election
-pub trait ElectionProviderBase {
-	/// The account identifier type.
+/// An [`ElectionDataProvider`] that exposes for an external entity to request a lock/unlock on
+/// updates in the election data.
+pub trait LockableElectionDataProvider: ElectionDataProvider {
+	fn set_lock() -> data_provider::Result<()>;
+	fn unlock();
+}
+
+/// Something that can compute the result of an election and pass it back to the caller in a paged
+/// way.
+pub trait ElectionProvider {
+	/// The account ID identifier;
 	type AccountId;
 
 	/// The block number type.
 	type BlockNumber;
 
-	/// The error type that is returned by the provider.
-	type Error: Debug;
+	/// The error type returned by the provider;
+	type Error: Debug + PartialEq;
 
-	/// The upper bound on election winners that can be returned.
-	///
-	/// # WARNING
-	///
-	/// when communicating with the data provider, one must ensure that
-	/// `DataProvider::desired_targets` returns a value less than this bound. An
-	/// implementation can chose to either return an error and/or sort and
-	/// truncate the output to meet this bound.
-	type MaxWinners: Get<u32>;
+	/// The maximum number of winners per page in results returned by this election provider.
+	type MaxWinnersPerPage: Get<u32>;
+
+	/// The maximum number of backers that a single page may have in results returned by this
+	/// election provider.
+	type MaxBackersPerWinner: Get<u32>;
+
+	/// The number of pages that this election provider supports.
+	type Pages: Get<PageIndex>;
 
 	/// The data provider of the election.
 	type DataProvider: ElectionDataProvider<
 		AccountId = Self::AccountId,
 		BlockNumber = Self::BlockNumber,
 	>;
 
+	/// Elect a new set of winners.
+	///
+	/// The result is returned in a target major format, namely as vector of supports.
+	///
+	/// This should be implemented as a self-weighing function. The implementor should register its
```

Comment:
I think we did like this before as well, but curious why do we need this to be self-weighting? Isn't it cleaner that it returns the weight and caller registers the weight with the system? The trait does not need to assume this is always called from system hooks right?

Reply:
That can be done yes, I don't have a strong opinion on either of the approaches. I like the idea of keeping the trait cleaner, but not very happy with the implicit assumption that `fn elect` is self-weighting.

Reply:
I opened this issue to keep track of this change. It should be easy and can be a mentor task https://github.com/paritytech/polkadot-sdk/issues/6136